### DISCUSSION TITLE
Not able to install on Windows

### DISCUSSION THREAD

**Discussion by maggieroot39** [sebalamm/KaMIS#21] [Q&A] [answered]
Hello, I recently found this project and was trying to install it on Windows. Is this project not yet supported on Windows? Or is just I'm missing some environment?

I get the following info when I run ```bash .\compile_withcmake.sh```.

I know the last several ```cp``` error follows from earlier errors. And ```make: *** No rule to make target `NCORES'.  Stop.``` is because 'makefile' is not generated in ```cmake ../``` (line 17), so the next command ```make -j $NCORES``` in line 18 can not be exectued. 

But I'm not sure how those ```Performing Test COMPILER_SUPPORTS_FUNROLL_LOOPS - Failed```  during ```cmake ../``` can be fixed, or whether it is relevant to this issue.

Many thanks for your help!

```
-- Building for: Visual Studio 16 2019
-- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19043.
-- The C compiler identification is MSVC 19.29.30133.0
-- The CXX compiler identification is MSVC 19.29.30133.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test COMPILER_SUPPORTS_FUNROLL_LOOPS
-- Performing Test COMPILER_SUPPORTS_FUNROLL_LOOPS - Failed
-- Performing Test COMPILER_SUPPORTS_FNOSTACKLIMITS
-- Performing Test COMPILER_SUPPORTS_FNOSTACKLIMITS - Failed
-- Performing Test COMPILER_SUPPORTS_MARCH_NATIVE
-- Performing Test COMPILER_SUPPORTS_MARCH_NATIVE - Failed
-- Performing Test COMPILER_SUPPORTS_FPERMISSIVE
-- Performing Test COMPILER_SUPPORTS_FPERMISSIVE - Failed
-- Performing Test COMPILER_SUPPORTS_NOUNUSED
-- Performing Test COMPILER_SUPPORTS_NOUNUSED - Failed
-- Performing Test COMPILER_SUPPORTS_NOUNUSEDRES
-- Performing Test COMPILER_SUPPORTS_NOUNUSEDRES - Failed
-- Found OpenMP_C: -openmp (found version "2.0")
-- Found OpenMP_CXX: -openmp (found version "2.0")
-- Found OpenMP: TRUE (found version "2.0")
-- OpenMP support detected
CMake Deprecation Warning at wmis/CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.

-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/82534/OneDrive/Desktop/KaMIS-master/build
make: *** No rule to make target `NCORES'.  Stop.
cp: cannot stat `./build/redumis': No such file or directory
cp: cannot stat `./build/graphchecker': No such file or directory
cp: cannot stat `./build/sort_adjacencies': No such file or directory
cp: cannot stat `./build/online_mis': No such file or directory
cp: cannot stat `./build/wmis/branch_reduce': No such file or directory
cp: cannot stat `./build/wmis/weighted_ls': No such file or directory
```

**Comment by schulzchristian** [comment]
Hi, 

I am currently on vacation. Will see what I can do next week.

Best
Christian

**Reply by maggieroot39** [reply]
Thank you! I'll continue to try to figure it out myself and look forward to hearing from you when you're back :) Enjoy your vacation!

**Comment by schulzchristian** [comment, marked as answer]
Thanks! As far as I remember we never compiled it on Windows. If you figure it out, I would be happy to get the details and like to share that here.

Best
Christian

**Reply by maggieroot39** [reply]
After consulting a friend, I give up on directly install KaMIS on Windows, but install it through Ubuntu.

1. In ```Control Panel\Programs``` select ```Turn Windows features on or off```, make sure 'Windows subsystem for Linux' is selected. Reboot if needed.
2. Download Ubuntu (directly in Microsoft Store) and locate KaMIS folder (mine is ```cd ../../mnt/c/ (followed by path in C:/)```).
3. Install required g++, openMP and cmake, excuteable can be easily found by searching 'install xxx on Ubuntu'.
4. When you are in the KaMIS folder, install KaMIS using ```./compile_withcmake.sh```. (If there are permission issues, use ```sudo ./compile_withcmake.sh``` or ```sudo bash ./compile_withcmake.sh``` instead.)

And to run the command, use for example ```(path)./graphchecker FILENAME```.

Best,
Xinyi

**Reply by maggieroot39** [reply]
Hi, I think I've successfully installed the program. 
Though I have a small question while using the 'redumis' method: what does "setrlimit returned result = -1" mean? (Is it because my graph is not reduced?)

Thanks,
Xinyi

**Reply by Hespian** [reply]
Hi Xinyi, 

this message probably comes from [here](https://github.com/KarlsruheMIS/KaMIS/blob/f99641d719594b4f7091898c9002a8dc762213d2/lib/mis/kernel/ParFastKer/fast_reductions/src/MaximumMatching.cpp#L63). It is from an external algorithm for Maximum Matching that is used as a subroutine of the reduction algorithm. 

A non-zero return value usually means that the operation failed. In this case, the stack size should be increased. I'm not sure if setrlimit works at all under windows. 

You could try just removing that call and hope that the stack is large enough for your inputs. As far as I can tell the program should just crash if the stack is too small. Or you could try increasing the stack limit in some other way.

Best
Demian

**Reply by maggieroot39** [reply]
Hi Demian,

Thanks! The other method 'online_mis' worked well. So I'll try as you suggested :)

Update: remove that call and re-install worked. I guess as you suggested, it might be because setrlimit does not work under windows - as it keeps return -1 even in small enough graphs (the 6-vertex example provided in the package).

Best,
Xinyi

**Reply by Hespian** [reply]
Hi Xinyi,

can you clarify what is returning -1? If you commented out the part around setrlimit, it shouldn't return anything ;)

Best,
Demian

**Reply by maggieroot39** [reply]
Hi Demian,

Ah sorry for not clearifying it - after commenting the few lines containing 'sterlimit', the algorithm worked as expected, and it did not return anything regarding `sterlimit`. 

I meant it keep returning `-1` **before** I commented the `sterlimit` part, even for very small graphs. (So, as you suggested, it was probably because `sterlimit` does not work on windows, but not becuase of the stack size.)

Best,
Xinyi
